### ISSUE TITLE
[rewrite:router] Race condition when `route.set` is called before `resolve` in the `onmatch` handler.

### ISSUE THREAD

**Issue by pygy** [lhorie/mithril.js#1267]
http://jsbin.com/jaqomofopo/edit?js,console

```JS
m.route(document.body, '/', {
  '/':{
    onmatch: function(vnode, resolve){
      setTimeout(resolve, 1000)
    },
    render: function(){console.log('/')}
  },
  
  '/2': {view: function(){
    console.log("/2")
  }}
})

setTimeout(function(){m.route.set('/2')},100)
```

Expected result: `'/2' is rendered, then the `resolve` on `'/'` is ignored since the route has been set to `'/2'` in the mean time.

Actual result: `'/2' is rendered, then  `'/'` when its resolution finishes.

Solution: It can be prevented by adding the code on lines 24, 33 and 39 [here](https://github.com/pygy/mithril.js/blob/8e4fabf0ed4378641f200abeb02d3d99a2f322cc/api/router.js).

Assign a unique ID to each resolution phase, and remember only the last one as `globalId`. All but the last `resolve` callbacks are thus expired.

**Comment by lhorie** [comment]
Fixed in 00d555b9f41c071f1c455d6ead9f5e1bfcac684a

**Comment by pygy** [comment]
Well spotted, I hadn't realized you could combine the id with the `resolved` flag.
